File: src/main/java/com/my/admin/app/commerce/service/ClientService.java
package com.my.admin.app.commerce.service;

import com.black.food.manager.dao.ClientDao;
import com.black.food.manager.exception.ApplicationException;
import com.black.food.manager.model.api.request.client.CreateClientRequest;
import com.black.food.manager.model.entity.Client;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.util.Optional;

@Service
public class ClientService {

    private static final Logger LOGGER = LoggerFactory.getLogger(ClientService.class);

    private ClientDao clientDao;

    @Transactional
    public Client findByDniOrElseSave(final CreateClientRequest createClientRequest) {
        Optional<Client> clientOptional = findByEmail(createClientRequest.getEmail());

        return clientOptional.orElseGet(() -> {
            Client client = new Client();
            client.setName(createClientRequest.getName());
            client.setLastName(createClientRequest.getLastName());
            client.setPhone(createClientRequest.getPhone());
            client.setEmail(createClientRequest.getEmail());

            return save(client);
        });
    }

    private Optional<Client> findByEmail(final String email) {
        return clientDao.findByEmail(email);
    }

    @Transactional
    private Client save(final Client client) {
        Client clientSaved;
        try {
            clientSaved = clientDao.save(client);
        } catch (Exception constraintViolationException) {
            var message = "Ocurrio un error al almacenar el cliente, verifica los datos.";
            LOGGER.error(message);
            LOGGER.error(constraintViolationException.getLocalizedMessage());
            throw new ApplicationException(message, HttpStatus.CONFLICT);
        }
        return clientSaved;
    }

    @Autowired
    public void setClientDao(final ClientDao clientDao) {
        this.clientDao = clientDao;
    }
}

File: src/main/kotlin/com/bk/reactive/app/commerce/admin/my/service/ClientService.kt
package com.bk.reactive.app.commerce.admin.my.service

import com.bk.reactive.app.commerce.admin.my.dao.ClientDAO
import com.bk.reactive.app.commerce.admin.my.exception.api.ApplicationException
import com.bk.reactive.app.commerce.admin.my.model.api.request.client.CreateClientRequest
import com.bk.reactive.app.commerce.admin.my.model.document.Client
import mu.KotlinLogging
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono

@Service
class ClientService(
    private val clientDAO: ClientDAO
) {
    private val logger = KotlinLogging.logger { }

    fun findByDniOrElseSave(createClientRequest: CreateClientRequest): Mono<Client> {
        return findByEmail(createClientRequest.email)
            .switchIfEmpty(
                save(
                    Client(
                        id = null,
                        name = createClientRequest.name,
                        lastName = createClientRequest.lastName,
                        phone = createClientRequest.phone,
                        email = createClientRequest.email,
                        dni = createClientRequest.dni
                    )
                )
            )
    }

    private fun save(client: Client): Mono<Client> {
        return clientDAO.save(client)
            .onErrorResume {
                val message = "Ocurrio un error al almacenar el cliente, verifica los datos."
                logger.error { message }
                logger.error { it.localizedMessage }
                Mono.error {
                    ApplicationException(
                        message = message,
                        code = HttpStatus.CONFLICT, cause = Throwable(message)
                    )
                }
            }
    }

    private fun findByEmail(email: String): Mono<Client> = clientDAO.findByEmail(email)
}